module/state_synchronization/execution_data_requester.go

```diff
@@ -0,0 +1,548 @@
+			e.missingBlocks.add(seal.BlockID)
+			continue
+		}
+
+		sealedHeights = append(sealedHeights, sealedBlock.Header.Height)
+
+		executionData, err := e.byBlockID(ctx, seal.BlockID)
+		if err != nil {
+			sealLogger.Error().Err(err).Msg("failed to get execution data for block")
+			e.missingBlocks.add(seal.BlockID)
+			continue
+		}
+
+		sealLogger.Debug().Msgf("fetched execution data for height %d", sealedBlock.Header.Height)
+		go e.onExecutionDataFetched(executionData)
+	}
+
+	e.missingBlocks.save(ctx)
+
+	// Finally, check if there are any heights we've missed. The set of seals in a block must be
+	// for consecutive heights, so sort the heights and then check for any that are missing before
+	// the first seal
+	if len(sealedHeights) == 0 {
+		return
+	}
+
+	sort.Slice(sealedHeights, func(i, j int) bool {
+		return sealedHeights[i] < sealedHeights[j]
+	})
+
+	err = e.addMissingBefore(sealedHeights[0])
+	if err != nil {
+		logger.Error().Err(err).Msg("failed to check for missing seals")
+	}
+
+	e.last.set(ctx, sealedHeights[len(sealedHeights)-1])
+}
+
+// addMissingBefore searches for blocks missed by the finalized block processor, and enqueues them
+// to be fetched
+func (e *ExecutionDataRequester) addMissingBefore(current uint64) error {
+	next := e.last.height() + 1
+	if next >= current {
+		// current is either the next block in the sequence, or one we've already seen
+		return nil
+	}
+
+	for next < current {
+		e.log.Debug().Msgf("next: %d, sealed: %d", next, current)
+		nextBlock, err := e.blocks.ByHeight(next)
+		if err != nil {
+			// TODO: handle this condition better. If this error is encountered, the blocks will not
+			//       be flagged as missing and reattempted until the next boot when the full check is
+			//       performed.
+			return fmt.Errorf("failed to get block for height %d: %w", next, err)
+		}
+		e.missingBlocks.add(nextBlock.ID())
+		next++
+	}
+
+	return nil
+}
+
+// byBlockID fetches the ExecutionData for a block by its ID
+func (e *ExecutionDataRequester) byBlockID(signalerCtx irrecoverable.SignalerContext, blockID flow.Identifier) (*ExecutionData, error) {
+	// Get the ExecutionResult, which contains the root CID for the execution data
+	result, err := e.results.ByBlockID(blockID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup execution result for block: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(signalerCtx, edFetchTimeout)
+	defer cancel()
+
+	// Fetch the ExecutionData for blockID from the blobstore. If it doesn't exist locally, it will
+	// be fetched from the network.
+	executionData, err := e.eds.Get(ctx, result.ExecutionDataID)
+
+	// Data is either corrupt or malformed. delete and redownload
+	var malformedDataError *MalformedDataError
+	var blobSizeLimitExceededError *BlobSizeLimitExceededError
+	if errors.Is(err, malformedDataError) || errors.Is(err, blobSizeLimitExceededError) {
+		cid := flow.FlowIDToCid(result.ExecutionDataID)
+		e.bs.DeleteBlob(signalerCtx, cid)
```

Comment:
since we're currently only consuming sealed blocks, there's no need for expirations.

In this case, I'm deleting a blob that's known to have incorrect data.